chalk-ir/src/lib.rs

```diff
@@ -569,17 +576,23 @@ impl<TF: TypeFamily> ParameterData<TF> {
 }
 
 #[derive(Clone, PartialEq, Eq, Hash, PartialOrd, Ord, Fold, HasTypeFamily)]
-pub struct ProjectionTy<TF: TypeFamily> {
-    pub associated_ty_id: AssocTypeId<TF>,
-    pub substitution: Substitution<TF>,
+pub enum AliasTy<TF: TypeFamily> {
+    Projection(ProjectionTy<TF>),
+    ImplTrait(ImplTraitId<TF>),
```

Comment:
wouldn't the impl Trait need to be parameterized similar to the projection? E.g. in a situation like `foo<T>() -> impl Trait`, the `impl Trait` is really parameterized by the `T`, right? See for example [this playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=5dc9ff4597b49fc3eb5d6ec40a9b5e6c).

(Now I'm wondering how `foo(impl Trait) -> impl Trait` works... does it rely on rustc desugaring the argument impl Trait to a type parameter?)

Reply:
(Thinking a bit more about that, probably yes -- the APIT from outside the function, and the RPIT from inside the function, can't really be represented using `ImplTrait` anyway...)

Reply:
Good call, I've added substitutions.